apollo-router/src/plugins/rhai/engine.rs

```diff
@@ -0,0 +1,1201 @@
+                        } else {
+                            obj.with_mut(|request| *request.supergraph_request.uri_mut() = uri);
+                            Ok(())
+                        }
+                    }
+                }
+            );
+        )*
+    };
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct RhaiService {
+    pub(super) scope: Arc<Mutex<Scope<'static>>>,
+    pub(super) service: ServiceStep,
+    pub(super) engine: Arc<Engine>,
+    pub(super) ast: AST,
+}
+
+impl Rhai {
+    pub(super) fn run_rhai_service(
+        &self,
+        function_name: &str,
+        subgraph: Option<&str>,
+        service: ServiceStep,
+        scope: Arc<Mutex<Scope<'static>>>,
+    ) -> Result<(), String> {
+        let block = self.block.load();
+        let rhai_service = RhaiService {
+            scope: scope.clone(),
+            service,
+            engine: block.engine.clone(),
+            ast: block.ast.clone(),
+        };
+        let mut guard = scope.lock().unwrap();
+        // Note: We don't use `process_error()` here, because this code executes in the context of
+        // the pipeline processing. We can't return an HTTP error, we can only return a boxed
+        // service which represents the next stage of the pipeline.
+        // We could have an error pipeline which always returns results, but that's a big
+        // change and one that requires more thought in the future.
+        match subgraph {
+            Some(name) => {
+                block
+                    .engine
+                    .call_fn(
+                        &mut guard,
+                        &block.ast,
+                        function_name,
+                        (rhai_service, name.to_string()),
+                    )
+                    .map_err(|err| err.to_string())?;
+            }
+            None => {
+                block
+                    .engine
+                    .call_fn(&mut guard, &block.ast, function_name, (rhai_service,))
+                    .map_err(|err| err.to_string())?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn new_rhai_engine(path: Option<PathBuf>, sdl: String, main: PathBuf) -> Engine {
+        let mut engine = Engine::new();
+        // If we pass in a path, use it to configure our engine
+        // with a FileModuleResolver which allows import to work
+        // in scripts.
+        if let Some(scripts) = path {
+            let resolver = FileModuleResolver::new_with_path(scripts);
+            engine.set_module_resolver(resolver);
+        }
+
+        // The macro call creates a Rhai module from the plugin module.
+        let mut module = exported_module!(router_plugin);
+        combine_with_exported_module!(&mut module, "header", router_header_map);
+        combine_with_exported_module!(&mut module, "json", router_json);
+        combine_with_exported_module!(&mut module, "context", router_context);
```

Comment:
If we ever did a version 2 of this, it would be nice to keep those various modules separate.

Reply:
I combine them because I want all of them to be available at the global level, but maybe I can do that when registering the modules instead

Reply:
I don't think there's anything wrong with what you've done. I was just lamenting the fact that we didn't have the individual modules in the first place.